docs/devel/luns.md

```diff
@@ -0,0 +1,82 @@
+   - (Optional) UI checkbox to the user, before starting the migration.
+   - The migration _must_ be a cold migration.
+2. An oVirt version with unsupported API. The [PR](https://github.com/oVirt/ovirt-engine/pull/559) provides LUN details is required.
+   - We may check for missing LUN details and skip the disk migration.
+   - (Optional) Have a mechanism to check the oVirt version for future work.
+3. Creating a `PersistentVolume` requires the file system type: `ext4` or `xfs` of the LUN, which we don't know.
+   - Might not be required for `Block` type.
+4. On non-local destination, the controller needs permission to create the `PersistentVolumes`.
+   - It may be enough to be added to the forklift-operator.
+
+Implementation
+----------------------------------------
+
+The controller gather data from source provider and saves them to the inventory.
+As part of the data gathering we will gather the LUN disks data. We have the VMs disk attachments where we follow to the disks.
+Within the disks API, if the _storage\_type_ is `LUN`, we will gather from the API endpoint `api/disks/<disk_id>` the size, address, IQN and the mapping of the LUN.
+Those will be found withing the `logical_unit` element of the `lun_storage`.
+On the migration plan a new steps will be added, creating a `PersistentVolume`(PV) and a `PersistentVolumeClaim`(PVC) per LUN disk.
+Both of them will contain the relevant data to use the LUN, the PVC will be bound the the PV by a unique label, `<vm_id>-<vm_name>-vol-<volume-number>`.
+When the VM specification created, the controller will add a LUN disk using the PVC that was created as [described by kubevirt](https://kubevirt.io/user-guide/virtual_machines/disks_and_volumes/#lun).
+
+Relevant validations will be performed to know we have the LUN details:
+
+- Allow `LUN` storage type on existing validation.
+- LUN details exist on the provider.
+- SCSI persistent reservation (exists), may required further warning when LUN is detected.
+- (Optional) Make sure the VM is in `down` on source provider(might be not required and be part of the execution plan).
+
+Execution Flow
+----------------------------------------
+
+1. The inventory will be filled based on the provider.
+   - On each disk, try to gather `lun_storage` data.
+2. Validation rules will be executed.
+3. Two steps will be added to the migration:
+   - `CreatePersistentVolumes`, which will create the relevant `PV` based on the `LUN` details.
+   - `CreatePersistentVolumeClaims`, which will create the relevant `PVC` and **bound** to the above `PV`.
+On each step, it first will check the existence of the CRs, when they won't be available, it will be created.
+4. When building the VM specifications, create the `LUN` disk and volume associated with the `PVC`.
+5. Once the migration ends successfully, detach the `LUN` disk from the source VM.
+6. If the migration failed and it archieved, remove any new CR that was created (PV/PVC).
```

Comment:
and also attach the lun back to the vm on the source system?

and what failures do you refer to here? let's say that the destination cluster doesn't have access to the lun (btw, there's no preliminary check for that, right?) - would we still detach the lun from the source? the downside is that the user may end up in a vm that cannot be started on the destination without further changes, and cannot start the vm on the source without attaching the lun back

Reply:
no, the idea was to detach the disk only when the migration succeed, until then it will be available on the source.
the failures i refer to is each problem that may occur in the migration process.
we don't check access to the lun on destination cluster. you are right about having a "succeed" migration when the destination have no access and we will remain bald from both sides (until manual action is executed).
a warning / another step of validation to check access + documentation?